module/zstd/zstd.c

```diff
@@ -0,0 +1,622 @@
+
+	z = (struct zstd_kmem *)zstd_mempool_alloc(zstd_mempool_cctx, nbytes);
+
+	if (!z) {
+		return (NULL);
+	}
+
+	return ((void*)z + (sizeof (struct zstd_kmem)));
+}
+
+/*
+ * Allocator for zstd decompression context using mempool_allocator with
+ * fallback to reserved memory if allocation fails
+ */
+static void *
+zstd_dctx_alloc(void *opaque __maybe_unused, size_t size)
+{
+	size_t nbytes = sizeof (struct zstd_kmem) + size;
+	struct zstd_kmem *z = NULL;
+	enum zstd_kmem_type type = ZSTD_KMEM_DEFAULT;
+
+	z = (struct zstd_kmem *)zstd_mempool_alloc(zstd_mempool_dctx, nbytes);
+	if (!z) {
+		/* Try harder, decompression shall not fail */
+		z = vmem_alloc(nbytes, KM_SLEEP);
+		if (z) {
+			z->pool = NULL;
+		}
+	} else {
+		return ((void*)z + (sizeof (struct zstd_kmem)));
+	}
+
+	/* Fallback if everything fails */
+	if (!z) {
+		/* Barrier since we only can handle it in a single thread */
+		mutex_enter(&zstd_dctx_fallback.barrier);
+		mutex_exit(&zstd_dctx_fallback.barrier);
+		mutex_enter(&zstd_dctx_fallback.barrier);
+		z = zstd_dctx_fallback.mem;
+		type = ZSTD_KMEM_DCTX;
+	}
+
+	/* Allocation should always be successful */
+	if (!z) {
+		return (NULL);
+	}
+
+	z->kmem_type = type;
+	z->kmem_size = nbytes;
+
+	return ((void*)z + (sizeof (struct zstd_kmem)));
+}
+
+/* Free allocated memory by its specific type */
+static void
+zstd_free(void *opaque __maybe_unused, void *ptr)
+{
+	struct zstd_kmem *z = (ptr - sizeof (struct zstd_kmem));
+	enum zstd_kmem_type type;
+
+	ASSERT3U(z->kmem_type, <, ZSTD_KMEM_COUNT);
+	ASSERT3U(z->kmem_type, >, ZSTD_KMEM_UNKNOWN);
+
+	type = z->kmem_type;
+	switch (type) {
+	case ZSTD_KMEM_DEFAULT:
+		kmem_free(z, z->kmem_size);
+		break;
+	case ZSTD_KMEM_POOL:
+		zstd_mempool_free(z);
+		break;
+	case ZSTD_KMEM_DCTX:
+		mutex_exit(&zstd_dctx_fallback.barrier);
+		break;
+	default:
+		break;
+	}
+}
+
+/* Allocate fallback memory to ensure safe decompression */
+static void __init
+create_fallback_mem(struct zstd_fallback_mem *mem, size_t size)
+{
+	mem->mem_size = size;
+	mem->mem = vmem_zalloc(mem->mem_size, KM_SLEEP);
+	mutex_init(&mem->barrier, NULL, MUTEX_DEFAULT, NULL);
+}
+
+/* Initialize memory pool barrier mutexes */
+static void __init
+zstd_mempool_init(void)
+{
+	int i;
+
+	zstd_mempool_cctx = (struct zstd_pool *)
+	    kmem_zalloc(ZSTD_POOL_MAX * sizeof (struct zstd_pool), KM_SLEEP);
+	zstd_mempool_dctx = (struct zstd_pool *)
+	    kmem_zalloc(ZSTD_POOL_MAX * sizeof (struct zstd_pool), KM_SLEEP);
+
+	for (i = 0; i < ZSTD_POOL_MAX; i++) {
+		mutex_init(&zstd_mempool_cctx[i].barrier, NULL,
+		    MUTEX_DEFAULT, NULL);
+		mutex_init(&zstd_mempool_dctx[i].barrier, NULL,
+		    MUTEX_DEFAULT, NULL);
+	}
+}
+
+/* Initialize zstd-related memory handling */
+static int __init
+zstd_meminit(void)
+{
+	zstd_mempool_init();
+
+	/* Estimate the size of the fallback decompression context */
+	create_fallback_mem(&zstd_dctx_fallback,
+	    P2ROUNDUP(ZSTD_estimateDCtxSize() + sizeof (struct zstd_kmem),
```

Comment:
this is a static fallback memory array. it will only be allocated once. the size is also very small (about 16kb). this fallback array is important. if memory allocation fails for some reason (out of memory) decompression will not work and the resulting data is invalid. so in case memory cannot be allocated this fallback memory is used with a single thread lock. while running stress tests on a memory constrained vm i was able to run in such a problem and developed this solution

Reply:
I completely agree with the need for a fallback decompression buffer.  My concern was specifically that the buffer was large, but it sounds like that's not the case at all.  If the buffers only 16k then we might as well always allocate it, but then let's convert the `vmem_zalloc()` to a `kmem_zalloc()` in `create_fallback_mem()`.  There's no reason to use the virtual address space for this.

Reply:
thats true. and it is always allocated already. but nobody can guarantee that the context wont increase its size to lets say 65kb in future. and keep in mind . the kmem_alloc function has a warning if its bigger than 64kb. 
i said 16kb, but this is not the exact value. the exact value is returned by a function provided by zstd api. i have to check the exact size. but its very small

Reply:
okay latest zstd version on X64 has a decompression context size of 160312 bytes. so its better to stay with vmem

Reply:
so, there's nothing to do, right?

Reply:
not for allocation. but for cleanup the barrier can be removed. i already have commited this in my PR

Reply:
Then let's add a comment with the expected maximum size to avoid any future confusion.  This memory also should be released with `vmem_free()` instead of `kmem_free()` in `zstd_fini()`.

Reply:
ack; will do that after the tests finished successfully. don't want to touch too much

Reply:
done b11e79